framework/lib/components/select/select-field.tsx

```diff
@@ -49,7 +49,7 @@ export function SelectField<T extends Option> ({
                     ? values.map((item: any) => item.value)
                     : values.value
                 )
-                onChangeCallback(values as T | T[], event)
+                onChangeCallback(values as T & T[], event)
```

Comment:
The type intersection `T & T[]` looks like an invalid type, as `T` and `T[]` usually don't have any common methods. Could you please explain the idea behind this / provide a code sample that that is supposed to work with?

Reply:
Furthermore, if you want to change the type based on `isMulti`, you should probably use generics.

Reply:
Currently if you declare e to be Option in the example below, the function is not assignable to the onChange prop because of mismatching types.
`<SelectField
                isClearable={ false }
                isMulti={ false }
                size="sm"
                aria-label="limit-dimensions-to"
                onChange={ (e: Option) => addDimensionsLimit('to', e.value) }
                value={
                  limitDimensions
                    ? {
                      value: limitPath.to,
                      label: capitalize(limitPath.to),
                    }
                    : { value: '', label: '' }
                }
                options={ [
                  { label: 'Largest', value: 'largest' },
                  { label: 'Smallest', value: 'smallest' },
                ] }
                name={ `dataSets.${name}.options.limit.to` }
                label=""
              />`

The idea is that if  isMulti is true you would change the function to be
 `
(e: Option[]) => something(map((item) => item.value, e))
`

Reply:
@olchyk98  Does having the SelectFieldProps like this make more sense?

```
export type SelectFieldProps<T> = Omit<SelectProps<T>, 'onChange'> &
InputFieldProps & {
  onChange?: ((val: T, event: ActionMeta<T>) => void) | ((val: T[], event: ActionMeta<T>) => void)
  direction?: StackDirection
  name: string
  label: string
  validate?: RegisterOptions
  isRequired?: boolean
  isMulti?: boolean
}
```

Reply:
While that may be better, this is not 100% clean. Returning "or" types in those cases requires another runtime check, otherwise you'll never know for sure if you're getting `Option[]`, or `Option`.

What I would suggest it to create a generic that would handle this behaviour, in this example generic `K`:

1. Adding another generic to the component, creates a simple declaration that can be used to link `isMulti` prop to `onChange` callback value. After a few small changes in props interface, we'll forward `K` to the props interface. 
```
export function SelectField<T extends Option, K extends boolean = true> ({
```

2. Replacing `isMulti` type from `boolean` to generic.
```
export type SelectFieldProps<T, K extends boolean = true> = ... & {
  onChange?: (val: K extends true ? T[] : T, event: ActionMeta<T>) => void
  isMulti?: K
}
``` 

Note, that we're always using `extends boolean = true`, It tells compiler that `isMulti` in that case, may be only a boolean (defaulted to true)

Reply:
I'll be in the office soon, let me know if you need any help with this!